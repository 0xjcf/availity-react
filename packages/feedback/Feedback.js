import React from 'react';
import PropTypes from 'prop-types';
import { UncontrolledDropdown, DropdownToggle, DropdownMenu } from 'reactstrap';
import FeedbackForm from './FeedbackForm';

const Feedback = ({
  appName,
  formProps,
  children,
  className,
  color,
  outline,
  prompt,
  onFeedbackSent,
  ...props
}) => (
  <UncontrolledDropdown className={`${className} hidden-print`} {...props}>
    <DropdownToggle color={color} outline={outline}>
      {children || 'Give Feedback'}
    </DropdownToggle>
    <DropdownMenu right className="p-3" style={{ width: '400px' }}>
      <FeedbackForm
        name={appName}
        onFeedbackSent={onFeedbackSent}
        prompt={prompt}
        {...formProps}
      />
    </DropdownMenu>
  </UncontrolledDropdown>
);

Feedback.propTypes = {
  appName: PropTypes.string.isRequired,
  className: PropTypes.string,
  prompt: PropTypes.string,
  color: PropTypes.string,
  outline: PropTypes.bool,
  formProps: PropTypes.object,
  children: PropTypes.node,
  onFeedbackSent: PropTypes.func,
};

Feedback.defaultProps = {
  className: '',
  color: 'light',
};

export default Feedback;
