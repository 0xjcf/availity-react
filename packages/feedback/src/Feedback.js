import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown, DropdownToggle } from 'reactstrap';
import { avLogMessagesApi } from '@availity/api-axios';
import { useToggle } from '@availity/hooks';
import FeedbackDropdown from './FeedbackDropdown';
import FeedbackModal from './FeedbackModal';

const Feedback = ({
  appName,
  modal,
  zIndex,
  supportZIndex,
  children,
  analytics,
  className,
  outline,
  color,
  formProps,
  prompt,
  onFeedbackSent,
  showSupport,
  onModalOpenChange,
  ...props
}) => {
  const [feedbackIsOpen, feedbackToggle] = useToggle(false);

  const handleFeedbackToggle = () => {
    feedbackToggle();
    onModalOpenChange(!feedbackIsOpen);
  };

  return (
    <Dropdown
      isOpen={feedbackIsOpen && !modal}
      toggle={handleFeedbackToggle}
      className={`${className} hidden-print`}
      {...props}
    >
      <DropdownToggle color={color} outline={outline}>
        {children || showSupport ? 'Feedback & Support' : 'Give Feedback'}
      </DropdownToggle>
      {modal ? (
        <FeedbackModal
          onFeedbackSent={onFeedbackSent}
          prompt={prompt}
          isOpen={feedbackIsOpen}
          zIndex={zIndex}
          toggle={handleFeedbackToggle}
          name={appName}
          analytics={analytics}
          showSupport={showSupport}
          supportZIndex={supportZIndex}
          feedbackToggle={feedbackToggle}
          {...formProps}
        />
      ) : (
        <FeedbackDropdown
          onFeedbackSent={onFeedbackSent}
          feedbackToggle={feedbackToggle}
          prompt={prompt}
          analytics={analytics}
          toggle={handleFeedbackToggle}
          name={appName}
          showSupport={showSupport}
          supportZIndex={supportZIndex}
          modal={modal}
          {...formProps}
        />
      )}
    </Dropdown>
  );
};

Feedback.propTypes = {
  appName: PropTypes.string,
  modal: PropTypes.bool,
  zIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  supportZIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  children: PropTypes.node,
  className: PropTypes.string,
  outline: PropTypes.bool,
  color: PropTypes.string,
  formProps: PropTypes.object,
  prompt: PropTypes.string,
  onFeedbackSent: PropTypes.func,
  onModalOpenChange: PropTypes.func,
  analytics: PropTypes.shape({
    info: PropTypes.func.isRequired,
  }),
  showSupport: PropTypes.bool,
};

Feedback.defaultProps = {
  modal: false,
  color: 'secondary',
  analytics: avLogMessagesApi,
  showSupport: false,
  onModalOpenChange: Object,
};

export default Feedback;
