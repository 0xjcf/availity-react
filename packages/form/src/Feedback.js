import React from 'react';
import PropTypes from 'prop-types';
import { FormFeedback } from 'reactstrap';
import { ErrorMessage } from 'formik';

const Feedback = ({ name, ...rest }) => {
  const feedbackId = `${name}-feedback`.toLowerCase();
  const AvFeedback = ({ children, ...rest }) => (
    <FormFeedback {...rest}>
      <i className="icon icon-attention" aria-label="Error" />
      {children}
    </FormFeedback>
  );

  AvFeedback.propTypes = {
    children: PropTypes.node,
  };
  return <ErrorMessage id={feedbackId} component={AvFeedback} name={name} {...rest} />;
};

Feedback.propTypes = {
  name: PropTypes.string.isRequired,
};

export default Feedback;
