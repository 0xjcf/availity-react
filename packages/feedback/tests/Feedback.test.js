import React from 'react';
import { render } from 'react-testing-library';
import Feedback from '..';

describe('Feedback', () => {
  test('should render', () => {
    const { container } = render(<Feedback appName="Payer Space" />);

    expect(container).toMatchSnapshot();
  });

  test('should render custom button text', () => {
    const { container } = render(
      <Feedback appName="Payer Space">Provide Valuable Feedback</Feedback>
    );

    expect(container).toMatchSnapshot();
  });

  test('should render custom button color', () => {
    const { container } = render(
      <Feedback appName="Payer Space" color="success" />
    );

    expect(container).toMatchSnapshot();
  });

  test('should render custom prompt text', () => {
    const { container } = render(
      <Feedback appName="Payer Space" prompt="Provide same feedback" />
    );

    expect(container).toMatchSnapshot();
  });

  test('should render custom form props', () => {
    const { container } = render(
      <Feedback
        appName="Payer Space"
        formProps={{
          aboutOptions: [
            { value: 'payerSpace', label: 'Payer Space' },
            { value: 'wholeSite', label: 'Whole Site' },
          ],
        }}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should render custom className', () => {
    const { container } = render(
      <Feedback appName="Payer Space" className="junk and-stuff" />
    );

    expect(container).toMatchSnapshot();
  });
});
