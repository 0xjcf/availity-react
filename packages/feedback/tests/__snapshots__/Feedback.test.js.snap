// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback should render 1`] = `
<div>
  <div
    class=" hidden-print dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-light"
      type="button"
    >
      Give Feedback
    </button>
    <div
      aria-hidden="true"
      class="p-3 dropdown-menu dropdown-menu-right"
      role="menu"
      style="width: 400px;"
      tabindex="-1"
    >
      <div
        class="text-center"
      >
        <p>
          Tell us what you think about Payer Space.
        </p>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-smile"
            />
            <span
              class="sr-only"
            >
              Smiley face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-meh"
            />
            <span
              class="sr-only"
            >
              Meh face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-frown"
            />
            <span
              class="sr-only"
            >
              Frowny face
            </span>
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Please select one of the faces above
        </div>
        <textarea
          aria-label="Please provide any feedback, requests, and/or defects."
          class="mt-3 form-control"
          placeholder="Feedback? Requests? Defects? (optional)"
        />
        <div
          class="mt-3 text-right"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feedback should render custom button color 1`] = `
<div>
  <div
    class=" hidden-print dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-success"
      type="button"
    >
      Give Feedback
    </button>
    <div
      aria-hidden="true"
      class="p-3 dropdown-menu dropdown-menu-right"
      role="menu"
      style="width: 400px;"
      tabindex="-1"
    >
      <div
        class="text-center"
      >
        <p>
          Tell us what you think about Payer Space.
        </p>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-smile"
            />
            <span
              class="sr-only"
            >
              Smiley face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-meh"
            />
            <span
              class="sr-only"
            >
              Meh face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-frown"
            />
            <span
              class="sr-only"
            >
              Frowny face
            </span>
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Please select one of the faces above
        </div>
        <textarea
          aria-label="Please provide any feedback, requests, and/or defects."
          class="mt-3 form-control"
          placeholder="Feedback? Requests? Defects? (optional)"
        />
        <div
          class="mt-3 text-right"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feedback should render custom button text 1`] = `
<div>
  <div
    class=" hidden-print dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-light"
      type="button"
    >
      Provide Valuable Feedback
    </button>
    <div
      aria-hidden="true"
      class="p-3 dropdown-menu dropdown-menu-right"
      role="menu"
      style="width: 400px;"
      tabindex="-1"
    >
      <div
        class="text-center"
      >
        <p>
          Tell us what you think about Payer Space.
        </p>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-smile"
            />
            <span
              class="sr-only"
            >
              Smiley face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-meh"
            />
            <span
              class="sr-only"
            >
              Meh face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-frown"
            />
            <span
              class="sr-only"
            >
              Frowny face
            </span>
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Please select one of the faces above
        </div>
        <textarea
          aria-label="Please provide any feedback, requests, and/or defects."
          class="mt-3 form-control"
          placeholder="Feedback? Requests? Defects? (optional)"
        />
        <div
          class="mt-3 text-right"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feedback should render custom className 1`] = `
<div>
  <div
    class="junk and-stuff hidden-print dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-light"
      type="button"
    >
      Give Feedback
    </button>
    <div
      aria-hidden="true"
      class="p-3 dropdown-menu dropdown-menu-right"
      role="menu"
      style="width: 400px;"
      tabindex="-1"
    >
      <div
        class="text-center"
      >
        <p>
          Tell us what you think about Payer Space.
        </p>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-smile"
            />
            <span
              class="sr-only"
            >
              Smiley face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-meh"
            />
            <span
              class="sr-only"
            >
              Meh face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-frown"
            />
            <span
              class="sr-only"
            >
              Frowny face
            </span>
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Please select one of the faces above
        </div>
        <textarea
          aria-label="Please provide any feedback, requests, and/or defects."
          class="mt-3 form-control"
          placeholder="Feedback? Requests? Defects? (optional)"
        />
        <div
          class="mt-3 text-right"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feedback should render custom form props 1`] = `
<div>
  <div
    class=" hidden-print dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-light"
      type="button"
    >
      Give Feedback
    </button>
    <div
      aria-hidden="true"
      class="p-3 dropdown-menu dropdown-menu-right"
      role="menu"
      style="width: 400px;"
      tabindex="-1"
    >
      <div
        class="text-center"
      >
        <p>
          Tell us what you think about Payer Space.
        </p>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-smile"
            />
            <span
              class="sr-only"
            >
              Smiley face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-meh"
            />
            <span
              class="sr-only"
            >
              Meh face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-frown"
            />
            <span
              class="sr-only"
            >
              Frowny face
            </span>
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Please select one of the faces above
        </div>
        <select
          class="custom-select"
          type="select"
        >
          <option
            value="payerSpace"
          >
            Payer Space
          </option>
          <option
            value="wholeSite"
          >
            Whole Site
          </option>
        </select>
        <textarea
          aria-label="Please provide any feedback, requests, and/or defects."
          class="mt-3 form-control"
          placeholder="Feedback? Requests? Defects? (optional)"
        />
        <div
          class="mt-3 text-right"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Feedback should render custom prompt text 1`] = `
<div>
  <div
    class=" hidden-print dropdown"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      class="btn btn-light"
      type="button"
    >
      Give Feedback
    </button>
    <div
      aria-hidden="true"
      class="p-3 dropdown-menu dropdown-menu-right"
      role="menu"
      style="width: 400px;"
      tabindex="-1"
    >
      <div
        class="text-center"
      >
        <p>
          Provide same feedback
        </p>
        <div
          class="btn-group"
          role="group"
        >
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-smile"
            />
            <span
              class="sr-only"
            >
              Smiley face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-meh"
            />
            <span
              class="sr-only"
            >
              Meh face
            </span>
          </button>
          <button
            class="btn btn-light"
            style="font-size: 1.4em; padding: .2em .4em;"
            type="button"
          >
            <span
              class="icon icon-frown"
            />
            <span
              class="sr-only"
            >
              Frowny face
            </span>
          </button>
        </div>
        <div
          class="invalid-feedback"
        >
          Please select one of the faces above
        </div>
        <textarea
          aria-label="Please provide any feedback, requests, and/or defects."
          class="mt-3 form-control"
          placeholder="Feedback? Requests? Defects? (optional)"
        />
        <div
          class="mt-3 text-right"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
